### DISCUSSION TITLE
Type checking for table classes?

### DISCUSSION THREAD

**Discussion by jenskutilek** [fonttools/fonttools#2859] [Q&A] [answered]
I'm using VSCode which does live type checking. In projects involving FontTools, it generates a lot of noise because it can't check the attributes of SFNT table classes.

E.g.

```py
glyf = font["glyf"]
print(sorted(glyf.keys())
```

```
Cannot access member keys for "DefaultTable": Member "keys" is unknown.
```

I've tried adding stubs for the TTFont object, but how can I define that specific keys return specific table classes?

```py
def __getitem__(self, tag: str) -> DefaultTable: ...
```

BTW I've added the stubs I have so far to this [new repo](https://github.com/jenskutilek/Python-Stubs).

**Comment by justvanrossum** [comment, marked as answer]
`TTFont` partially behaves like a dict, so I'd expect we should do something like `typing.TypedDict`, but I'm not sure how to do that for a class implementing the (Mutable)Mapping protocol.

If it's just about suppressing the noise, perhaps use `typing.Any` instead of `DefaultTable`.
